File: src/main/java/frc/robot/subsystems/DrivetrainFalcon.java
package frc.robot.subsystems;

import com.ctre.phoenix.motorcontrol.ControlMode;
import com.ctre.phoenix6.controls.Follower;
import com.ctre.phoenix6.signals.NeutralModeValue;
import com.pathplanner.lib.auto.AutoBuilder;
import com.pathplanner.lib.util.ReplanningConfig;
import edu.wpi.first.math.controller.SimpleMotorFeedforward;
import edu.wpi.first.math.geometry.Pose2d;
import edu.wpi.first.math.geometry.Rotation2d;
import edu.wpi.first.math.kinematics.ChassisSpeeds;
import edu.wpi.first.math.kinematics.DifferentialDriveOdometry;
import edu.wpi.first.math.kinematics.DifferentialDriveWheelSpeeds;
import edu.wpi.first.networktables.NetworkTableEntry;
import edu.wpi.first.wpilibj.DriverStation;
import edu.wpi.first.wpilibj.RobotBase;
import edu.wpi.first.wpilibj.smartdashboard.Field2d;
import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
import edu.wpi.first.wpilibj2.command.SubsystemBase;
import frc.robot.Constants;
import frc.robot.util.DifferentialDriveDebug;
import frc.robot.util.NavX;
import frc.robot.util.PIDControllerDebug;
import frc.robot.util.SafeTalonFX;

/**
 * Initializes the drivetrain using falcon500
 */
public class DrivetrainFalcon extends SubsystemBase {

    public static Field2d m_fieldSim = new Field2d();
    // These values are used for Autonomous
    private final double ticks_per_wheel_revolution = 26112.0;

    // NEED THIS FOR AUTONOMOUS
    // private val max_ticks_per_hundred_milliseconds: Double = ticks_per_foot * Constants.MAX_SPEED_LOW_GEAR / 10
    private final double ticks_per_foot = ticks_per_wheel_revolution / (Constants.WHEEL_DIAMETER_FT * Math.PI); // .8 inches is diameter of wheel in feet
    // Initializing Master Falcon Motors
    private final SafeTalonFX leftMasterFalcon = new SafeTalonFX(Constants.LEFT_MASTER_PORT, true, false); // change to false for no PID?
    private final SafeTalonFX rightMasterFalcon = new SafeTalonFX(Constants.RIGHT_MASTER_PORT, true, false);
    // Initializing Slave Falcon Motors
    private final SafeTalonFX leftSlaveFalcon = new SafeTalonFX(Constants.LEFT_SLAVE_PORT, true, false);

    //    private val leftGroup = MotorControllerGroup(leftMasterFalcon, leftSlaveFalcon)
//    private val rightGroup = MotorControllerGroup(rightMasterFalcon, rightSlaveFalcon)
    private final SafeTalonFX rightSlaveFalcon = new SafeTalonFX(Constants.RIGHT_SLAVE_PORT, true, false);
    //    private val kinematics = DifferentialDriveKinematics(Constants.TRACK_WIDTH)
    private final DifferentialDriveDebug diffDrive = new DifferentialDriveDebug(leftMasterFalcon, rightMasterFalcon);
    private final NavX imu = new NavX();
    private final PIDControllerDebug leftPIDController = new PIDControllerDebug(0.02, 0.0, 0.0);
    private final PIDControllerDebug rightPIDController = new PIDControllerDebug(0.02, 0.0, 0.0);
    // TODO: ks and kv values need to be determined for the robot
    private final SimpleMotorFeedforward feedforward = new SimpleMotorFeedforward(1.0, 3.0);
    private final double isFullSpeed = 1.0;
    private final DifferentialDriveOdometry m_odometry = new DifferentialDriveOdometry(imu.getRotation2d(), 0.0, 0.0);
    private int loopIdx = 0;

    public DrivetrainFalcon() {

        resetAll();

        configurePathPlanner();

        if (RobotBase.isReal()) {

            rightMasterFalcon.setInverted(true);
            rightSlaveFalcon.setInverted(true);
            leftMasterFalcon.setInverted(false);
            leftSlaveFalcon.setInverted(false);

            leftSlaveFalcon.setControl(new Follower(leftMasterFalcon.getDeviceID(), false));
            rightSlaveFalcon.setControl(new Follower(rightMasterFalcon.getDeviceID(), false));

            setNeutralMode(NeutralModeValue.Brake);

//             leftMasterFalcon.configClosedloopRamp(Constants.DRIVETRAIN_RAMP);
            // rightMasterFalcon.configClosedloopRamp(Constants.DRIVETRAIN_RAMP);

            // orchestra = new Orchestra();
            // orchestra.addInstrument(leftMasterFalcon);
            // orchestra.addInstrument(rightMasterFalcon);
            // orchestra.addInstrument(leftSlaveFalcon);
            // orchestra.addInstrument(rightSlaveFalcon);

            // orchestra.loadMusic("test.chrp");
//            diffDrive.setDeadband(0.02)
        }

        // diffDrive.setRightSideInverted(false);
        //diffDrive.isSafetyEnabled = false; //WARN:this line of kotlin commented out because I don't believe it does anything useful

        // shifter.shiftLow();

        configurePID();

//        resetEncoders();



        SmartDashboard.putData("Field", m_fieldSim);
    }

    private void resetAll() {
        var pose = new Pose2d();
        resetPose(pose);
    }

    private void configurePathPlanner() {
        AutoBuilder.configureRamsete(
                m_odometry::getPoseMeters,
                this::resetPose,
                this::getCurrentChassisSpeeds,
                this::curvatureDriveChassisSpeeds,
                // b?,
                // zeta?,
                new ReplanningConfig(), // Default path replanning config. See the API for the options here
                () -> {
                    // Boolean supplier that controls when the path will be mirrored for the red alliance
                    // This will flip the path being followed to the red side of the field.
                    // THE ORIGIN WILL REMAIN ON THE BLUE SIDE
                    var alliance = DriverStation.getAlliance();
                    if (alliance.isPresent()) {
                        return alliance.get() == DriverStation.Alliance.Red;
                    }
                    return false;
                },
                this // Reference to this subsystem to set requirements
        );
    }

    public double getIsFullSpeed() {
        return isFullSpeed;
    }

    private void setNeutralMode(NeutralModeValue coastMode) {
        rightMasterFalcon.setNeutralMode(coastMode);
        leftMasterFalcon.setNeutralMode(coastMode);
    }

    public double getAverageDist() {
        double leftDist = leftMasterFalcon.getPosition().getValue() / ticks_per_foot;
        double rightDist = rightMasterFalcon.getPosition().getValue() / ticks_per_foot;
        return (leftDist + rightDist) * 0.5;
    }

    public void resetPose(Pose2d pose) {
        System.out.println("INFO: DrivetrainFalcon.resetOdometry");
        resetEncoders();
        // Ideally these should both be zero now.
        var leftDist = leftMasterFalcon.getPosition().getValue();
        var rightDist = leftMasterFalcon.getPosition().getValue();

        var alliance = DriverStation.getAlliance();
        if (alliance.isPresent()) {
            // We are using Blue-alliance always field orientation.
            if (alliance.get() == DriverStation.Alliance.Red) {
                imu.setAngleOffset(180.0);
            } else {
                imu.setAngleOffset(0.0);
            }
        }

        m_odometry.resetPosition(imu.getRotation2d(), leftDist, rightDist, pose);
    }

    public void configurePID() {
        // Set Velocity PID Constants in slot 0
//        leftMasterFalcon.config_kF(0, Constants.LEFT_VELOCITY_FF);
//        leftMasterFalcon.config_kP(0, Constants.LEFT_VELOCITY_P);
//        leftMasterFalcon.config_kI(0, Constants.LEFT_VELOCITY_I);
//        leftMasterFalcon.config_kD(0, Constants.LEFT_VELOCITY_D);
//
//        rightMasterFalcon.config_kF(0, Constants.RIGHT_VELOCITY_FF);
//        rightMasterFalcon.config_kP(0, Constants.RIGHT_VELOCITY_P);
//        rightMasterFalcon.config_kI(0, Constants.RIGHT_VELOCITY_I);
//        rightMasterFalcon.config_kD(0, Constants.RIGHT_VELOCITY_D);
    }

    public void configureMotionMagic() {
//        leftMasterFalcon.configMotionCruiseVelocity(Constants.LEFT_MASTER_VELOCITY, Constants.kTIMEOUT_MS);
//        leftMasterFalcon.configMotionAcceleration(Constants.LEFT_MASTER_ACCELERATION, Constants.kTIMEOUT_MS);
//
//        rightMasterFalcon.configMotionCruiseVelocity(Constants.RIGHT_MASTER_VELOCITY, Constants.kTIMEOUT_MS);
//        rightMasterFalcon.configMotionAcceleration(Constants.RIGHT_MASTER_ACCELERATION, Constants.kTIMEOUT_MS);
    }

    private void debugPeriodic() {
        loopIdx++;
        if (loopIdx == 10) {
            loopIdx = 0;

            if (RobotBase.isReal()) {
//                SmartDashboard.putNumber("L

//                SmartDashboard.putNumber("Heading", calcHeading())

//                leftCurrent.setNumber(leftMasterFalcon.statorCurrent)
                // leftPosition!!.setNumber(leftMasterFalcon.selectedSensorPosition)
//                 leftVelocity!!.setNumber(leftMasterFalcon.selectedSensorVelocity)

//                rightCurrent.setNumber(rightMasterFalcon.statorCurrent)
                // rightPosition!!.setNumber(rightMasterFalcon.selectedSensorPosition)
//                 rightVelocity!!.setNumber(rightMasterFalcon.selectedSensorVelocity)


            } else {
                double curLeftCurrent = 0.0;

                // if (simIter.hasNext()) {
                //   curLeftCurrent = simIter.next();
                // }
//                leftCurrent.setNumber(curLeftCurrent)
//                leftPosition.setNumber(0.0)
//                leftVelocity.setNumber(0.0)
//
//                rightCurrent.setNumber(0.0)
//                rightPosition.setNumber(0.0)
//                rightVelocity.setNumber(0.0)

            }
        }
    }

    private void curvatureDriveChassisSpeeds(ChassisSpeeds chassisSpeeds) {
        double speed = chassisSpeeds.vxMetersPerSecond / Constants.MAX_METERS_PER_SECOND_VELOCITY;
        double rotation = chassisSpeeds.omegaRadiansPerSecond / Constants.MAX_ANGULAR_VELOCITY;
        curvatureDrive(speed, rotation);
    }

    private void curvatureDrive(double speed, double rotation) {
        curvatureDrive(speed, rotation, false);
    }

    public void curvatureDrive(double speed, double rotation, boolean isQuickTurn) {
        diffDrive.curvatureDrive(speed, rotation, isQuickTurn);
    }

    public void resetEncoders() {
        System.out.println("INFO: DrivetrainFalcon.resetEncoders");
        leftMasterFalcon.setPosition(0.0);
        rightMasterFalcon.setPosition(0.0);
    }

//    public double calcDist() {
//        double x = m_odometry.getPoseMeters().getX();
//        double y = m_odometry.getPoseMeters().getY();
//        double dist = Math.hypot(x, y);
//        return dist;
//    }

//    public double calcHeading() {
//        double x = m_odometry.getPoseMeters().getX();
//        double y = m_odometry.getPoseMeters().getY();
////        val delta = ((m_odometry.poseMeters.rotation.degrees % 360) + 360) % 360 - 180 // delta % 360 is to set the input between -360 and 360
//        double atanDegree = Math.toDegrees(Math.atan2(y, x));
//        return atanDegree;
//    }

//    public void setSpeeds(DifferentialDriveWheelSpeeds speeds) {
//        double leftFeedforward = feedforward.calculate(speeds.leftMetersPerSecond);
//        double rightFeedforward = feedforward.calculate(speeds.rightMetersPerSecond);
//        double leftOutput =
//                leftPIDController.calculate(leftMasterFalcon.getPosition().getValue(), speeds.leftMetersPerSecond);
//        double rightOutput =
//                rightPIDController.calculate(rightMasterFalcon.getPosition().getValue(), speeds.rightMetersPerSecond);
////        leftGroup.setVoltage(leftOutput + leftFeedforward)
////        rightGroup.setVoltage(rightOutput + rightFeedforward)
//    }

//    public void drive(double xSpeed, double rotation) {
////        setSpeeds(kinematics.toWheelSpeeds(ChassisSpeeds(xSpeed, 0.0, rotation)))
//        diffDrive.arcadeDrive(xSpeed, rotation);
//    }

//    public double[] getPositions() {
//        double[] pos = new double[2];
//        pos[0] = leftMasterFalcon.getPosition().getValue();
//        pos[1] = rightMasterFalcon.getPosition().getValue();
//        return pos;
//    }

    public double[] getVelocities() {
        double[] velocites = new double[2];
        velocites[0] = leftMasterFalcon.getVelocity().getValue();
        velocites[1] = rightMasterFalcon.getVelocity().getValue();
        return velocites;
    }

    public DifferentialDriveWheelSpeeds getDifferentialDriveWheelSpeeds() {
        return new DifferentialDriveWheelSpeeds(m_leftEncoder.getRate(), m_rightEncoder.getRate());
    }

    public ChassisSpeeds getCurrentChassisSpeeds() {
        var diffWheelSpeeds = getDifferentialDriveWheelSpeeds();
        return Constants.DriveConstants.kDriveKinematics.toChassisSpeeds(difWheelSpeeds);
    }

//    public void set(double leftPercent, double rightPercent) {
//        leftMasterFalcon.set(leftPercent);
//        rightMasterFalcon.set(rightPercent);
//    }

//    public void set(ControlMode controlMode, double leftMagnitude, double rightMagnitude) {
//        leftMasterFalcon.set(leftMagnitude);
//        rightMasterFalcon.set(rightMagnitude);
//    }

    public void stop() {
        leftMasterFalcon.stopMotor();
        rightMasterFalcon.stopMotor();
    }

    private void updateOdometry() {
        Rotation2d rotation2d = imu.getRotation2d();
        double leftDist = leftMasterFalcon.getPosition().getValue();
        double rightDist = rightMasterFalcon.getPosition().getValue();

        m_odometry.update(rotation2d, leftDist, rightDist);
    }

    @Override
    public void periodic() {
        updateOdometry();
        m_fieldSim.setRobotPose(m_odometry.getPoseMeters());
        debugPeriodic();
    }
}

File: src/main/java/frc/robot/subsystems/DrivetrainFalcon.kt
package frc.robot.subsystems

import com.ctre.phoenix.motorcontrol.ControlMode
import com.ctre.phoenix.motorcontrol.InvertType
import com.ctre.phoenix.motorcontrol.NeutralMode
import com.ctre.phoenix.sensors.WPI_Pigeon2
import edu.wpi.first.math.controller.SimpleMotorFeedforward
import edu.wpi.first.math.geometry.Pose2d
import edu.wpi.first.math.geometry.Rotation2d
import edu.wpi.first.math.kinematics.DifferentialDriveOdometry
import edu.wpi.first.math.kinematics.DifferentialDriveWheelSpeeds
import edu.wpi.first.networktables.NetworkTableEntry
import edu.wpi.first.wpilibj.RobotBase
import edu.wpi.first.wpilibj.shuffleboard.BuiltInWidgets
import edu.wpi.first.wpilibj.shuffleboard.Shuffleboard
import edu.wpi.first.wpilibj.smartdashboard.Field2d
import edu.wpi.first.wpilibj2.command.SubsystemBase
import frc.robot.Constants
import frc.robot.util.*
import kotlin.math.atan2
import kotlin.math.hypot

/**
 * Initializes the drivetrain using falcon500
 */
class DrivetrainFalcon : SubsystemBase() {

    // These values are used for Autonomous
//    private val ticks_per_wheel_revolution = 26112.0
//    private val ticks_per_foot = ticks_per_wheel_revolution / (Constants.WHEEL_DIAMETER_FT * Math.PI) // .8 inches is diameter of wheel in feet

    // NEED THIS FOR AUTONOMOUS
    // private val max_ticks_per_hundred_milliseconds: Double = ticks_per_foot * Constants.MAX_SPEED_LOW_GEAR / 10

    // Initializing Master Falcon Motors
    private val usePID = true

    private val leftMasterFalcon = SafeTalonFX(Constants.LEFT_MASTER_PORT, isDrivetrain=true, usePID=usePID)
    private val rightMasterFalcon = SafeTalonFX(Constants.RIGHT_MASTER_PORT, isDrivetrain=true, usePID=usePID)
    private val leftSlaveFalcon = SafeTalonFX(Constants.LEFT_SLAVE_PORT, isDrivetrain=true, usePID=usePID)
    private val rightSlaveFalcon = SafeTalonFX(Constants.RIGHT_SLAVE_PORT, isDrivetrain=true, usePID=usePID)

//    private val leftGroup = MotorControllerGroup(leftMasterFalcon, leftSlaveFalcon)
//    private val rightGroup = MotorControllerGroup(rightMasterFalcon, rightSlaveFalcon)

    private val diffDrive = DifferentialDriveDebug(leftMasterFalcon, rightMasterFalcon)
    private val imu = NavX()

    private val leftPIDController = PIDControllerDebug(0.02, 0.0, 0.0)
    private val rightPIDController = PIDControllerDebug(0.02, 0.0, 0.0)

    // TODO: ks and kv values need to be determined for the robot
    private val feedforward = SimpleMotorFeedforward(1.0, 3.0)

    lateinit var leftCurrent: NetworkTableEntry
    lateinit var leftPosition: NetworkTableEntry
    lateinit var leftVelocity: NetworkTableEntry

    lateinit var rightCurrent: NetworkTableEntry
    lateinit var rightPosition: NetworkTableEntry
    lateinit var rightVelocity: NetworkTableEntry

    var isFullSpeed = 1.0

    private var loopIdx = 0

    private val m_odometry = DifferentialDriveOdometry(Rotation2d())

    private lateinit var pidgey: WPI_Pigeon2

    private lateinit var driveSim: DrivebaseSimFX

    enum class CoastMode {
        Coast, Brake
    }

    val dff: (Double) -> String = TalonFXUtil.df::format

    init {
        leftSlaveFalcon.follow(leftMasterFalcon)
        rightSlaveFalcon.follow(rightMasterFalcon)

        rightMasterFalcon.inverted = true
        rightSlaveFalcon.setInverted(InvertType.FollowMaster)
        leftMasterFalcon.inverted = false
        leftSlaveFalcon.setInverted(InvertType.FollowMaster)

        setCoastMode(CoastMode.Brake)

        if (RobotBase.isReal()) {

            // leftMasterFalcon.configClosedloopRamp(Constants.DRIVETRAIN_RAMP);
            // rightMasterFalcon.configClosedloopRamp(Constants.DRIVETRAIN_RAMP);

            // orchestra = new Orchestra();
            // orchestra.addInstrument(leftMasterFalcon);
            // orchestra.addInstrument(rightMasterFalcon);
            // orchestra.addInstrument(leftSlaveFalcon);
            // orchestra.addInstrument(rightSlaveFalcon);

            // orchestra.loadMusic("test.chrp");
//            diffDrive.setDeadband(0.02)
        } else {
            pidgey = WPI_Pigeon2(1)
            driveSim = DrivebaseSimFX(leftMasterFalcon, rightMasterFalcon, pidgey, m_odometry)
        }

        // diffDrive.setRightSideInverted(false);
        diffDrive.isSafetyEnabled = false

        // shifter.shiftLow();

         configurePID();
        // configureMotionMagic();

//        configureSmartDashBoard()

        resetEncoders()

//        SmartDashboard.putData("Field", m_fieldSim)
    }


    private fun setCoastMode(coastMode: CoastMode) {
        // Sets neutralMode to Coast or Brake depending on coastMode
        var neutralMode = NeutralMode.Coast // By default this is Coast
        if(coastMode == CoastMode.Coast) {
            neutralMode = NeutralMode.Coast
        } else if (coastMode == CoastMode.Brake) {
            neutralMode = NeutralMode.Brake
        }
        rightMasterFalcon.setNeutralMode(neutralMode)
        rightSlaveFalcon.setNeutralMode(neutralMode)
        leftMasterFalcon.setNeutralMode(neutralMode)
        leftSlaveFalcon.setNeutralMode(neutralMode)
    }

    fun getAverageDist() : Double {
        val leftDist = TalonFXUtil.nativeUnitsToDistanceFeet(leftMasterFalcon.selectedSensorPosition)
        val rightDist = TalonFXUtil.nativeUnitsToDistanceFeet(rightMasterFalcon.selectedSensorPosition)
        return (leftDist + rightDist) * 0.5
    }

    private fun updateOdometry() {
        println("INFO: updateOdometry")
//        var leftDist = TalonFXUtil.nativeUnitsToDistanceFeet(leftMasterFalcon.selectedSensorPosition)
//        var rightDist = TalonFXUtil.nativeUnitsToDistanceFeet(rightMasterFalcon.selectedSensorPosition)
        var leftDist = TalonFXUtil.nativeUnitsToDistanceMeters(leftMasterFalcon.selectedSensorPosition)
        var rightDist =TalonFXUtil. nativeUnitsToDistanceMeters(rightMasterFalcon.selectedSensorPosition)

        // Simulation values not yet available from NavX2. Get from the Pigeon IMU instead.
        var rotation2d = if (RobotBase.isReal()) imu.rotation2d else pidgey.rotation2d

        println("INFO: Gyro: ${dff(rotation2d.degrees)}, Left: ${dff(leftDist)}, Right: ${dff(rightDist)}")

        m_odometry.update(rotation2d, leftDist, rightDist)
    }

    fun resetOdometry(pose: Pose2d) {
        leftMasterFalcon.selectedSensorPosition = 0.0
        rightMasterFalcon.selectedSensorPosition = 0.0

        m_odometry.resetPosition(pose, imu.rotation2d)
    }

    fun resetGyro() { imu.reset() }

    fun getPose(): Pose2d { return m_odometry.poseMeters }

    fun configurePID() {

        // Set Velocity PID Constants in slot 0
        leftMasterFalcon.config_kF(0, Constants.LEFT_VELOCITY_FF)
        leftMasterFalcon.config_kP(0, Constants.LEFT_VELOCITY_P)
        leftMasterFalcon.config_kI(0, Constants.LEFT_VELOCITY_I)
        leftMasterFalcon.config_kD(0, Constants.LEFT_VELOCITY_D)

        rightMasterFalcon.config_kF(0, Constants.RIGHT_VELOCITY_FF)
        rightMasterFalcon.config_kP(0, Constants.RIGHT_VELOCITY_P)
        rightMasterFalcon.config_kI(0, Constants.RIGHT_VELOCITY_I)
        rightMasterFalcon.config_kD(0, Constants.RIGHT_VELOCITY_D)

    }

    fun configureMotionMagic() {
        leftMasterFalcon.configMotionCruiseVelocity(Constants.LEFT_MASTER_VELOCITY, Constants.kTIMEOUT_MS)
        leftMasterFalcon.configMotionAcceleration(Constants.LEFT_MASTER_ACCELERATION, Constants.kTIMEOUT_MS)

        rightMasterFalcon.configMotionCruiseVelocity(Constants.RIGHT_MASTER_VELOCITY, Constants.kTIMEOUT_MS)
        rightMasterFalcon.configMotionAcceleration(Constants.RIGHT_MASTER_ACCELERATION, Constants.kTIMEOUT_MS)
    }

    fun configureSmartDashBoard() {
        leftCurrent = generateEntry("Left Current", 0, 0)
        leftPosition = generateEntry("Left Position", 2, 0)
        leftVelocity = generateEntry("Left Velocity", 4, 0)
        rightCurrent = generateEntry("Right Current", 0, 2)
        rightPosition = generateEntry("Right Position", 2, 2)
        rightVelocity = generateEntry("Right Velocity", 4, 2)
    }

    override fun simulationPeriodic() {
        driveSim.run()
    }

    override fun periodic() {
        updateOdometry()

//        println("DrivetrainFalcon periodic")
//        SmartDashboard.putNumber("Encoder Ticks - Left", leftMasterFalcon.selectedSensorPosition)
//        SmartDashboard.putNumber("Encoder Ticks - Right", rightMasterFalcon.selectedSensorPosition)
//        SmartDashboard.putNumber(
//            "Encoder Rate (Normalized) - Left",
//            leftMasterFalcon.selectedSensorVelocity / max_ticks_per_hundred_milliseconds
//        )
//        SmartDashboard.putNumber(
//            "Encoder Rate (Normalized) - Right",
//            rightMasterFalcon.selectedSensorVelocity / max_ticks_per_hundred_milliseconds
//        )
//
//        SmartDashboard.putNumber("NavX Angle", gyro.rotation2d.degrees)
//
//        SmartDashboard.putNumber("Right Master Current", rightMasterFalcon.statorCurrent)
//        SmartDashboard.putNumber("Right Slave Current", rightSlaveFalcon.statorCurrent)
//        SmartDashboard.putNumber("Left Master Current", leftMasterFalcon.statorCurrent)
//        SmartDashboard.putNumber("Left Slave Current", leftSlaveFalcon.statorCurrent)

        loopIdx++
        if (loopIdx == 10) {
            loopIdx = 0

            if (RobotBase.isReal()) {
//                SmartDashboard.putNumber("Left Encoder", leftMasterFalcon.selectedSensorPosition)
//                SmartDashboard.putNumber("Right Encoder", rightMasterFalcon.selectedSensorPosition)
//                SmartDashboard.putNumber("Left Drive Speed", leftMasterFalcon.selectedSensorVelocity / 22000)
//                SmartDashboard.putNumber("Right Drive Speed", rightMasterFalcon.selectedSensorVelocity / 22000)
//                SmartDashboard.putNumber("Left Drive Stator Current", leftMasterFalcon.statorCurrent)
//                SmartDashboard.putNumber("Right Drive Stator Current", rightMasterFalcon.statorCurrent)
//                SmartDashboard.putNumber("Left Drive Supply Current", leftMasterFalcon.supplyCurrent)
//                SmartDashboard.putNumber("Right Drive Supply Current", rightMasterFalcon.supplyCurrent)

//                SmartDashboard.putNumber("Heading", calcHeading())

//                leftCurrent.setNumber(leftMasterFalcon.statorCurrent)
                // leftPosition!!.setNumber(leftMasterFalcon.selectedSensorPosition)
//                 leftVelocity!!.setNumber(leftMasterFalcon.selectedSensorVelocity)

//                rightCurrent.setNumber(rightMasterFalcon.statorCurrent)
                // rightPosition!!.setNumber(rightMasterFalcon.selectedSensorPosition)
//                 rightVelocity!!.setNumber(rightMasterFalcon.selectedSensorVelocity)


            } else {
                var curLeftCurrent = 0.0

                // if (simIter.hasNext()) {
                //   curLeftCurrent = simIter.next();
                // }
//                leftCurrent.setNumber(curLeftCurrent)
//                leftPosition.setNumber(0.0)
//                leftVelocity.setNumber(0.0)
//
//                rightCurrent.setNumber(0.0)
//                rightPosition.setNumber(0.0)
//                rightVelocity.setNumber(0.0)

            }
        }
    }

    fun arcadeDrive(speed: Double, rotation: Double, squareInputs: Boolean) {
        diffDrive.arcadeDrive(speed, rotation, squareInputs)
    }


    fun curvatureDrive(speed: Double, rotation: Double, isQuickTurn: Boolean) {
        diffDrive.curvatureDrive(speed, rotation, isQuickTurn)
    }

    /**
     * Sets the encoder values back to zero
     */
    fun resetEncoders() {
        println("Reset Encoders called")
        leftMasterFalcon.selectedSensorPosition = 0.0
        rightMasterFalcon.selectedSensorPosition = 0.0
    }

    fun calcDist(): Double {
        val x = m_odometry.poseMeters.x
        val y = m_odometry.poseMeters.y

        val dist = hypot(x, y)

        return dist
    }
    /**
     * @return atanDegree: The heading the robot needs to face towards the goal
     */
    fun calcHeading(): Double {
        val x = m_odometry.poseMeters.x
        val y = m_odometry.poseMeters.y
//        val delta = ((m_odometry.poseMeters.rotation.degrees % 360) + 360) % 360 - 180 // delta % 360 is to set the input between -360 and 360
        val atanDegree = Math.toDegrees(atan2(y, x))

//        println("INFO: X Position: $x, Y Position: $y, Delta: $delta, Heading: $atanDegree")

        return atanDegree
    }

    fun setSpeeds(speeds: DifferentialDriveWheelSpeeds) {
        val leftFeedforward = feedforward.calculate(speeds.leftMetersPerSecond)
        val rightFeedforward = feedforward.calculate(speeds.rightMetersPerSecond)
        val leftOutput =
            leftPIDController.calculate(leftMasterFalcon.selectedSensorPosition, speeds.leftMetersPerSecond)
        val rightOutput =
            rightPIDController.calculate(rightMasterFalcon.selectedSensorPosition, speeds.rightMetersPerSecond)

//        leftGroup.setVoltage(leftOutput + leftFeedforward)
//        rightGroup.setVoltage(rightOutput + rightFeedforward)

    }

    /**
     * Controls the robot using arcade drive.
     *
     * @param xSpeed: Double / The speed for the X axis
     * @param rotation: Double / The rotation
     */
    fun drive(xSpeed: Double, rotation: Double) {
//        setSpeeds(kinematics.toWheelSpeeds(ChassisSpeeds(xSpeed, 0.0, rotation)))
        diffDrive.arcadeDrive(xSpeed, rotation)
    }

    /**
     * @return double array of positions [left, right]
     */
    fun getPositions(): DoubleArray {
        return doubleArrayOf(leftMasterFalcon.selectedSensorPosition, rightMasterFalcon.selectedSensorPosition)
    }

    /**
     * @return double array of velocities [left, right]
     */
    fun getVelocities(): DoubleArray {
        return doubleArrayOf(leftMasterFalcon.selectedSensorVelocity, rightMasterFalcon.selectedSensorVelocity)
    }

    /**
     * Sets the left and right motors to a percent output
     * @param leftPercent Double
     * @param rightPercent Double
     */
    fun set(leftPercent: Double, rightPercent: Double) {
        leftMasterFalcon.set(leftPercent)
        rightMasterFalcon.set(rightPercent)
    }

    fun set(controlMode: ControlMode, leftMagnitude: Double, rightMagnitude: Double) {
        leftMasterFalcon.set(controlMode, leftMagnitude)
        rightMasterFalcon.set(controlMode, rightMagnitude)
    }

    /**
     * Immediately stops the drivetrain, only use in emergencies
     */
    fun stop() {
        leftMasterFalcon.stopMotor()
        rightMasterFalcon.stopMotor()
    }

    /**
     * Helper function to generate NetworkTableEntries
     */
    private fun generateEntry(entryName: String, columnIndex: Int, rowIndex: Int): NetworkTableEntry {
        return Shuffleboard.getTab("Drivetrain")
            .add(entryName, 0)
            .withSize(2, 2)
            .withPosition(columnIndex, rowIndex)
            .withWidget(BuiltInWidgets.kGraph)
            .entry
    }

    fun driveLeft(value: Double) { leftMasterFalcon.set(value) }

    fun driveRight(value: Double) { rightMasterFalcon.set(value) }
}